daemon/openrazer_daemon/misc/battery_notifier.py

```diff
@@ -72,26 +72,27 @@ def notify_battery(self):
         now = datetime.datetime.now()
 
         if (now - self._last_notify_time).seconds > self.frequency:
-            # Update last notified
-            self._last_notify_time = now
-
             battery_level = self._get_battery_func()
-            battery_percent = int(round(battery_level, 0))
-
-            # Sometimes on wifi don't get batt
-            if battery_level == -1.0:
-                time.sleep(0.2)
-                battery_level = self._get_battery_func()
+            battery_percent = round(battery_level)
+
+            # Sometimes due to various issues we don't get the percentage correctly.
+            # Just ignore them and don't show a bogus notification.
+            # See also: https://github.com/openrazer/openrazer/issues/2122
+            if battery_level in (0.0, -1.0):
+                self._logger.debug("Got bogus battery value: {0}, ignoring.".format(battery_level))
+                # Since we don't update _last_notify_time here we're going to retry very soon again.
+                # Sleep a bit so we don't spam the device with requests.
+                time.sleep(10)
```

Comment:
You don't need to sleep her.  Just return and the caller will sleep for you.

Reply:
Sleep for 0.1 seconds, if some device consistently returns 0 for some reason I don't want to spam CPU and USB bus because of that.

Reply:
Maybe you didn't notice, but I moved the update of `_last_notify_time`, so it won't get spammed?

Reply:
What do you have `self.frequency` set to?  I thought `self.frequency` was supposed to control how often you collect requests?

Reply:
If you look at the code in this file it's a bit funky written. There's a loop that calls notify_battery every 0.1 seconds and only that function decides it's not time yet to call. No clue why it was written that way (the code dates back to 2016), and it should probably get refactored because it's already now a pretty tight loop it seems.

I'm going to merge the current patch now, feel free to improve the code in this function, it's definitely desperately needed! And I personally don't really care about battery notifier since I don't own any wireless Razer device.

Reply:
I agree, the code is super weird.  It would have been better to just sleep by `self.frequency` in the first place.

Thanks for the merge!

Just want to say though that if you don't remove `sleep(10)` , someone may complain that setting the `frequency` parameter seems to have no effect sometimes.  Essentially 10 becomes the temporary frequency when it hits bad data.  And if it hits  a lot of bad data in a row, you may not get an update in quite a while.  Also, is 10Hz really spamming anyway?  Seems like video games would poll the mouse more often than that.

Reply:
> Just want to say though that if you don't remove sleep(10) , someone may complain that setting the frequency parameter seems to have no effect sometimes.

I'm aware of that and will put the blame on myself if that happens ;) Hopefully once I have some time to improve the driver we'll have less (or no) of these situations where the driver randomly reports 0 because of missing error handling.

> Also, is 10Hz really spamming anyway? Seems like video games would poll the mouse more often than that.

The daemon is running always, video games only when you actually want to run them. So I'm more careful there.